discover.py

```diff
@@ -0,0 +1,258 @@
+import re
+import logging
+import time
+import shutil
+from threading import Thread
+from datetime import datetime
+
+from zeroconf import ServiceBrowser, Zeroconf
+import netifaces
+import flask
+
+import supervisor_client
+
+log = logging.getLogger('discovery')
+
+SERVICE_TYPE = "_liquid._tcp.local."
+HOST_ONLY_NETWORK_MASK = '255.255.255.255'
+DNSMASQ_PROCESS_NAME = 'dnsmasq-dns'
+
+nodes = {}
+
+def reload_dnsmasq():
+    log.debug("Reloading dnsmasq-dns")
+    supervisor = supervisor_client.connect()
+    dnsmasq_process_info = supervisor.getProcessInfo(DNSMASQ_PROCESS_NAME)
+    if dnsmasq_process_info['statename'] == 'RUNNING':
+        supervisor.stopProcess(DNSMASQ_PROCESS_NAME)
+    supervisor.startProcess(DNSMASQ_PROCESS_NAME)
+
+def stop_dnsmasq():
+    log.debug("Stopping dnsmasq-dns")
+    supervisor = supervisor_client.connect()
+    supervisor.stopProcess(DNSMASQ_PROCESS_NAME)
+
+def rewrite_dnsmasq_conf(path, conf_string):
+    temp = path + ".tmp"
+    with open(temp, 'w') as f:
+        f.write(conf_string)
+    shutil.move(temp, path)
+
+def manage_dnsmasq():
+    def update_dnsmasq():
+        nonlocal last_dnsmasq_conf_string
+        interface = app.config['DNS_INTERFACE']
+        if not interface:
+            if last_dnsmasq_conf_string:
+                stop_dnsmasq()
+                last_dnsmasq_conf_string = ''
+            return
+
+        conf_string = get_dns_conf_string(interface)
+        if conf_string == last_dnsmasq_conf_string:
+            return
+
+        rewrite_dnsmasq_conf(app.config['DNS_CONF_FILE'], conf_string)
+        reload_dnsmasq()
+        last_dnsmasq_conf_string = conf_string
```

Comment:
I think we should update `last_dnsmasq_conf_string` *before* calling out to supervisor. In case the supervisor command fails, there is no sense in retrying that, every 3 seconds.